x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -289,10 +289,16 @@ private void handleIndexActionAuthorizationResult(final IndexAuthorizationResult
         final Authentication authentication = requestInfo.getAuthentication();
         final TransportRequest request = requestInfo.getRequest();
         final String action = requestInfo.getAction();
-        if (result.getIndicesAccessControl() != null) {
-            putTransientIfNonExisting(AuthorizationServiceField.INDICES_PERMISSIONS_KEY,
-                result.getIndicesAccessControl());
+        IndicesAccessControl indicesAccessControl = result.getIndicesAccessControl();
+        if (indicesAccessControl != null) {
+            IndicesAccessControl existing = threadContext.getTransient(AuthorizationServiceField.INDICES_PERMISSIONS_KEY);
+            if (existing == null) {
+                threadContext.putTransient(AuthorizationServiceField.INDICES_PERMISSIONS_KEY, indicesAccessControl);
+            } else {
+                existing.addPermissionsIfNotPresent(indicesAccessControl);
+            }
         }
```

Comment:
Conceptually I would not oppose the "narrowing" of privileges as proposed above, but I have two objections:
- `IndicesAccessControl.ALLOW_NO_INDICES` and `IndicesAccessControl.ALLOW_ALL` are dubiously similar, for DLS/FLS. When you add to `ALLOW_ALL` you restrict it, but when you add to `ALLOW_NO_INDICES` you also restrict it? I think I don't fully comprehend how `ALLOW_NO_INDICES` works for DLS .
- We're combining DLS/FLS permissions for different action names, i.e. there might be an index action that triggers  a different index action (async search) . What set of DLS permissions should the later have?